src/bifrost.cpp

```diff
@@ -0,0 +1,276 @@
+
+
+    opt.k = kmer;
+    opt.nb_threads = threads;
+    opt.verbose = verb;
+    opt.prefixFilenameOut = output_prefix;
+
+    std::string filename;
+    if (is_ref && (infile_2 == "NA")) {
+        while (std::getline(infile1, filename))
+        {
+            opt.filename_ref_in.push_back(filename);
+        }
+    } else if (!is_ref && (infile_2 == "NA"))
+    {
+        while (std::getline(infile1, filename))
+        {
+            opt.filename_seq_in.push_back(filename);
+        }
+    } else {
+        while (std::getline(infile1, filename))
+        {
+            opt.filename_ref_in.push_back(filename);
+        }
+        while (std::getline(infile2, filename))
+        {
+            opt.filename_seq_in.push_back(filename);
+        }
+    }
+
+    ColoredCDBG<> ccdbg(opt.k);
+    ccdbg.buildGraph(opt);
+    ccdbg.simplify(opt.deleteIsolated, opt.clipTips, opt.verbose);
+    ccdbg.buildColors(opt);
+
+    if (write_graph)
+    {
+        ccdbg.write(opt.prefixFilenameOut, opt.nb_threads, opt.verbose);
+    }
+
+    return ccdbg;
+}
+
+// generate colours for a unitig
+template <class T, class U, bool is_const>
+std::vector<bool> generate_colours(const UnitigMap<DataAccessor<T>, DataStorage<U>, is_const> unitig,
+                                   const size_t nb_colours,
+                                   const size_t position,
+                                   const bool single_mapping)
+{
+    // get colours information for unitig
+    const auto colourset = unitig.getData()->getUnitigColors(unitig);
+    std::vector<bool> colours_arr(nb_colours, 0);
+
+    // initialise a iterator, will only determine colours of single kmer, have to do for head and tail as may be different
+    UnitigColors::const_iterator it_uc = colourset->begin(unitig);
+    UnitigColors::const_iterator it_uc_end = colourset->end();
+
+    // if not single mapping, look at specific position
+    if (!single_mapping)
+    {
+        // iterate over specified kmer colours, update colours array with presence of those colours
+        for (it_uc; it_uc != it_uc_end; it_uc++)
+        {
+            if (it_uc.getKmerPosition() == position)
+            {
+                colours_arr[it_uc.getColorID()] = 1;
+            }
+        }
+    }
+    // else, look at only position in unitig given by mapping, ignore position
+    else {
+        // iterate over specified kmer colours, update colours array with presence of those colours
+        for (it_uc; it_uc != it_uc_end; it_uc++)
+        {
+            colours_arr[it_uc.getColorID()] = 1;
+        }
+    }
+    return colours_arr;
+}
+
+// negate colour arrays
+std::vector<bool> negate_colours_array(const std::vector<bool>& array1, const std::vector<bool>& array2)
+{
+    std::vector<bool> output_array = array1;
+    for (size_t i = 0; i < array1.size(); i++)
+    {
+        if (array1[i] == 1 && array2[i] == 0)
+        {
+            output_array[i] = 0;
+        }
+    }
+    return output_array;

```

Comment:
Not suggesting that you change this here, but if using `std::vector<bool>` turns out to be a pinch-point in ggcaller you could have a look at bitmagic: http://bitmagic.io/apis.html